tests/xdc/history_replication_dlq_test.go

```diff
@@ -0,0 +1,454 @@
+		if err == io.EOF { // Don't need errors.Is because io.EOF should never be wrapped.
+			return replicationTasks
+		}
+		s.NoError(err)
+		replicationTasks = append(replicationTasks, task)
+	}
+}
+
+// Execute the replication task as-normal, but also send it to the channel so that the test can wait for it to
+// know that the namespace data has been replicated.
+func (t *testNamespaceReplicationTaskExecutor) Execute(
+	ctx context.Context,
+	task *replicationspb.NamespaceTaskAttributes,
+) error {
+	err := t.replicationTaskExecutor.Execute(ctx, task)
+	if err != nil {
+		return err
+	}
+	select {
+	case t.tasks <- task:
+	default:
+	}
+	return nil
+}
+
+// PutReplicationTaskToDLQ is the same as the normal execution manager, but also sends the request to the channel so
+// that the test can wait for it to know that the replication task has been added to the DLQ.
+func (t *testExecutionManager) PutReplicationTaskToDLQ(
+	ctx context.Context,
+	request *persistence.PutReplicationTaskToDLQRequest,
+) error {
+	err := t.ExecutionManager.PutReplicationTaskToDLQ(ctx, request)
+	if err != nil {
+		return err
+	}
+	select {
+	case t.dlqRequests <- request:
+	default:
+	}
+	return nil
+}
+
+// Execute the replication task as-normal or return an error if the workflow ID matches the one that we want to fail.
+func (f testReplicationTaskExecutor) Execute(
+	ctx context.Context,
+	replicationTask *replicationspb.ReplicationTask,
+	forceApply bool,
+) error {
+	if replicationTask.GetHistoryTaskAttributes().WorkflowId == workflowIDToFail {
+		select {
+		case f.tasks <- replicationTask:
+		default:
+		}
+		return errors.New("failed to apply replication task")
+	}
+	return f.taskExecutor.Execute(ctx, replicationTask, forceApply)
+}
+
+// Convert the replication tasks using the base converter, but then wrap them in our own faulty executable tasks.
+func (t *testExecutableTaskConverter) Convert(
+	taskClusterName string,
+	clientShardKey replication.ClusterShardKey,
+	serverShardKey replication.ClusterShardKey,
+	replicationTasks ...*replicationspb.ReplicationTask,
+) []replication.TrackableExecutableTask {
+	convertedTasks := t.converter.Convert(taskClusterName, clientShardKey, serverShardKey, replicationTasks...)
+	testExecutableTasks := make([]replication.TrackableExecutableTask, len(convertedTasks))
+	for i, task := range convertedTasks {
+		testExecutableTasks[i] = &testExecutableTask{
+			replicationTaskExecutorParams: t.replicationTaskExecutorParams,
+			TrackableExecutableTask:       task,
+		}
+	}
+	return testExecutableTasks
+}
+
+// Execute the replication task as-normal or return an error if the workflow ID matches the one that we want to fail.
+func (t *testExecutableTask) Execute() error {
+	if et, ok := t.TrackableExecutableTask.(*replication.ExecutableHistoryTask); ok {
+		if et.WorkflowID == workflowIDToFail {
+			t.replicationTaskExecutorParams.tasks <- struct{}{}
+			return serviceerror.NewInvalidArgument("failed to apply replication task")
```

Comment:
I just wonder why there's a need to write code to return 2 errors when encountering `workflowIDToFail`? Naively I only expected there to be one.

Reply:
There are two different code paths for replication task processing (based on if streaming is enabled). One of them uses this testExecutableTask, and the other one uses testReplicationTaskExecutor. The test is trying to cover both cases I think.

Reply:
^ Yeah, that's why. I added a comment clarifying this.